### PR TITLE
Add in recursive abilities to chmod

### PR DESCRIPTION
**Pull request by morris25**
I thought it might be useful to have chmod change the permissions of subfiles

### CODE REVIEW

File: base/file.jl [hunk 1]
```diff
@@ -436,8 +436,13 @@ function readlink(path::AbstractString)
...
 
-function chmod(p::AbstractString, mode::Integer)
-    err = ccall(:jl_fs_chmod, Int32, (Cstring, Cint), p, mode)
+function chmod(path::AbstractString, mode::Integer; recursive::Bool=false)
+    if recursive && isdir(path)
+        for p in readdir(path)
+            chmod(joinpath(path, p), mode, recursive=true)
+        end
+    end
+    err = ccall(:jl_fs_chmod, Int32, (Cstring, Cint), path, mode)
```
Comment (StefanKarpinski) [inline base/file.jl]: I feel like you might want to change the directory's mode first and then recurse, no? That way if you don't have permission, you'll bail out sooner.

Comment (eschnett) [inline base/file.jl]: I often need to do `chmod -R u+w` on directory trees where I don't have write permission. This indeed only works if the permission is set first.

However, what about the converse, `chmod -R u-w`?

Comment (vtjnash) [inline base/file.jl]: i guess that would depend on whether you are adding or removing yourself from the execute (directory search) permission list?

File: doc/stdlib/file.rst [hunk 2]
```diff
@@ -75,11 +75,11 @@
 
    Returns the value of a symbolic link ``path``\ .
 
-.. function:: chmod(path, mode)
+.. function:: chmod(path, mode; recursive=false)
```
Comment (tkelman) [inline doc/stdlib/file.rst]: these changes also need to be made in the .jl docstring

**Comment by morris25** [comment]
I fixed the docstring and switched the recursion around. I also edited the test for windows.

File: base/file.jl [hunk 3]
```diff
@@ -436,8 +436,13 @@ function readlink(path::AbstractString)
     end
 end
 
-function chmod(p::AbstractString, mode::Integer)
-    err = ccall(:jl_fs_chmod, Int32, (Cstring, Cint), p, mode)
+function chmod(path::AbstractString, mode::Integer; recursive::Bool=false)
+    err = ccall(:jl_fs_chmod, Int32, (Cstring, Cint), path, mode)
+    if recursive && isdir(path)
```
Comment (tkelman) [inline base/file.jl]: this should probably also do some specific checking about symbolic links, take care about whether this follows through them

**Comment by morris25** [comment]
It now ignores links while recursing

**Comment by tkelman** [comment]
Does that match the coreutils behavior? Are the permissions of a symlink always the same as the permissions of the file it links to? Could use some tests for the symlink case (maybe unix-only, we can make junctions to directories but symlinks to files on windows requires non-default permissions).

**Comment by morris25** [comment]
It should match the coreutils behavior as described in [coreutils](http://www.gnu.org/software/coreutils/manual/html_node/chmod-invocation.html) as :


> for each symbolic link listed on the command line, chmod changes the permissions of the pointed-to file. In contrast, chmod ignores symbolic links encountered during recursive directory traversals.

 Although it differs from my own experiences. If this is wrong, please let me know and I'll fix it. I have added tests to ensure that any symlinked files are not changed during a recursive traversal.

**Comment by morris25** [comment]
I've rebased and cleaned up the tmpdir. Thanks for the pointers!

File: base/docs/helpdb/Base.jl [hunk 4]
```diff
@@ -5367,9 +5367,11 @@ is no file path involved, no path processing or fetching from node 1 is done.
 include_string
 
 """
-    chmod(path, mode)
+    chmod(path, mode; recursive=false)
 
-Change the permissions mode of `path` to `mode`. Only integer `mode`s (e.g. `0o777`) are currently supported.
+Change the permissions mode of `path` to `mode`. Only integer `mode`s (e.g. 0o777) are
```
Comment (omus) [inline base/docs/helpdb/Base.jl]: `0o777` should be in backticks.

File: test/file.jl [hunk 5]
```diff
@@ -63,6 +64,42 @@ chmod(file, filemode(file) & 0o7555)
...
+    chmod(dir, permissions, recursive=true)
+    @test filemode(dir) & 0o777 == permissions
+    @test filemode(subdir) & 0o777 == permissions
+    @test filemode(file) & 0o777 == permissions
+    @windows_only chmod(dir, 0o666, recursive=true)  # in case someone wants to use these later
+end
+@unix_only begin
+    mktempdir() do tmpdir
+        tmpfile=joinpath(tmpdir, "tempfile.txt")
+        close(open(tmpfile, "w"))
```
Comment (omus) [inline test/file.jl]: Can be simplified to `touch(tmpfile)`

**Comment by morris25** [comment]
Fixed!

File: test/file.jl [hunk 6]
```diff
@@ -63,6 +64,42 @@ chmod(file, filemode(file) & 0o7555)
...
+@windows_only begin
+    permissions = 0o444
+    @test filemode(dir) & 0o777 != permissions
+    @test filemode(subdir) & 0o777 != permissions
+    @test filemode(file) & 0o777 != permissions
+    chmod(dir, permissions, recursive=true)
+    @test filemode(dir) & 0o777 == permissions
+    @test filemode(subdir) & 0o777 == permissions
+    @test filemode(file) & 0o777 == permissions
+    @windows_only chmod(dir, 0o666, recursive=true)  # Reset permissions in case someone wants to use these later
```
Comment (tkelman) [inline test/file.jl]: `@windows_only` redundant here given the block?

**Comment by morris25** [comment]
Thanks, I must've missed that!

**Comment by morris25** [comment]
Is there anything else I should change?

**Comment by StefanKarpinski** [comment]
Looks good to me, but I'll let @tkelman have a last look. Thanks for reminding us to look here.

File: base/file.jl [hunk 7]
```diff
@@ -436,8 +436,15 @@ function readlink(path::AbstractString)
...
+        for p in readdir(path)
+            if !islink(joinpath(path, p))
+                chmod(joinpath(path, p), mode, recursive=true)
+            end
+        end
+    end
+    uv_error("chmod", err)
```
Comment (tkelman) [inline base/file.jl]: should the error be before the recursion?

**Comment by morris25** [comment]
I've changed the location of the err

**Comment by tkelman** [comment]
travis failure was #14991, unrelated
```
Exception running test subarray :
On worker 4:
LoadError: BoundsError: attempt to access (1,)
  at index [2]
 [inlined code] from ./abstractarray.jl:150
 in test_cartesian at /tmp/julia/share/julia/test/subarray.jl:96
 in runtests at /tmp/julia/share/julia/test/subarray.jl:241
 in runviews at /tmp/julia/share/julia/test/subarray.jl:249
 [inlined code] from /tmp/julia/share/julia/test/subarray.jl:328
```

**Comment by StefanKarpinski** [comment]
Thanks for seeing this through, @morris25!

**Pull request merged**